### PR TITLE
Lunchboxes Update (Sprites, Tweak, Bugfix)

### PR DESCRIPTION
**Pull request by Optimism333**
## Some Bite-Size Changes ##
- Lunchboxes are much more forgiving with items that they will hold, and are now able to store all the same things a small box can.
- Added four collectible lunchbox designs that didn't make it into the initial pr (thanks again to cricket in the Discord).
- Ablative lunchboxes will protect your lunch from lasers! They will not protect YOU from lasers, however.
- Addresses #34962

image

### Food Slices Had No Nutriment (Bugfix) ###
- Most food slice items inherited all their nutriment from the food item they were sliced from, but had no actual nutriment value if spawned on their own. So slices that spawned in lunchboxes wouldn't give the poor spaceman who bought them any nutritional value, and just disappear when an attempt was made to consume them.
- I added a check to fix this. After being spawned, lunchbox food items will check if they are empty of any reagents, and add some nutriment if they are. The amount of nutriment added depends on if the item is a main or snack/sweet.

[sprites] [tweak] [bugfix]

:cl:
 * rscadd: Added four new collectible lunchbox designs
 * rscadd: Ablative lunchboxes will protect your lunch from lasers!
 * tweak: Lunchboxes are much less restrictive about what items they can hold
 * bugfix: Lunchbox spawned food slice items will now have some nutritional value

### CODE REVIEW

**Comment by pedr69** [comment]
lunchbox_amogus

**Comment by SonixApache** [comment]
do #34962 while you're at it

**Comment by Eneocho** [comment]
> lunchbox_amogus

Only if it comes with an amogus flash toy inside.

**Comment by Optimism333** [comment]
So I've been told by an informant of a serious problem with lunchboxes. A fair few of the item list options are sliced food items, and sliced food items in our code (for the most part), have no nutriment content coded into them.

I need to make subtypes that actually do have nutriment and put those in the lunchbox lists instead, or many spessmen who buy lunchboxes will be in for some serious disappointment.

To-Fix:
- All sliced foods are fucked and have no nutriment
- The donut subtype I used has no nutriment
- Cherry jelly sandwich has woefully little nutriment, should use the subtype that at least also contains some cherry jelly

**Comment by Eneocho** [comment]
Yeah. Sliced whatevers actually inherit the nutrient from the parent upon slicing. A cheese wheel has 20 nutriment, slices into 5 pieces, each piece spawns with 4 nutriment.

**Comment by Optimism333** [comment]
Gonna convert this to a draft until that problem has been adequately dealt with.

**Comment by Eneocho** [comment]
Instead of making a subtype, I feel it would be better to give the slice itself the correct amount of chems inside by default and overwrite that when spawning a slice via cutting.
Something like: If food is not cut (spawned directly), add reagents, else, inherit reagents from parent source.
It would be better to maintain that way than having an extra copy that has reagents inside for each lunchbox item.
Then again, having a copy with reagents inside for lunchbox purposes would be the easier implementation.

**Comment by Optimism333** [comment]
> Instead of making a subtype, I feel it would be better to give the slice itself the correct amount of chems inside by default and overwrite that when spawning a slice via cutting. Something like: If food is not cut (spawned directly), add reagents, else, inherit reagents from parent source. It would be better to maintain that way than having an extra copy that has reagents inside for each lunchbox item. Then again, having a copy with reagents inside for lunchbox purposes would be the easier implementation.

You're probably right. Tomorrow or another day of this week, I will see if I can figure out a better solution to this problem. I will probably ask a collab for some feedback as well.

**Comment by pedr69** [comment]
does this only happen to slicables or basically all the food? i remember a while back banana creampies from maint also had 0 reagents

**Comment by Eneocho** [comment]
> does this only happen to slicables or basically all the food? i remember a while back banana creampies from maint also had 0 reagents

Only sliceables. Sliceables check how many slices it would spawn upon being sliced and spread its reagent contents equally among all the slices. Slices themselves are empty unless created via this process. An admin SPAWN for a bread slice would not have reagents. I vaguely remember the pie issue but that was a non-related issue as far as I am aware.

**Comment by Optimism333** [comment]
> > does this only happen to slicables or basically all the food? i remember a while back banana creampies from maint also had 0 reagents
> 
> Only sliceables. Sliceables check how many slices it would spawn upon being sliced and spread its reagent contents equally among all the slices. Slices themselves are empty unless created via this process. An admin SPAWN for a bread slice would not have reagents. I vaguely remember the pie issue but that was a non-related issue as far as I am aware.

What Eneocho said, yeah. I have a tentative solution in the works.

**Comment by Optimism333** [comment]
Okay, a fix of sorts has been applied. It seems to work as intended. There's probably some slice item edge cases that I missed, but hopefully not too many.

File: code/modules/reagents/reagent_containers/food/snacks.dm [hunk 1]
```diff
@@ -3513,6 +3513,12 @@
 	food_flags = FOOD_MEAT | FOOD_ANIMAL | FOOD_LACTOSE | FOOD_DIPPABLE
 	plate_offset_y = -4
 
+/obj/item/weapon/reagent_containers/food/snacks/meatbreadslice/New()
+	..()
+	spawn(5) // So that this isn't applied at the same time as the food slicing proc above that adds nutriment
```
Comment (pedr69) [inline code/modules/reagents/reagent_containers/food/snacks.dm:3518]: you don't need the spawns here

**Comment by Eneocho** [comment]
Not how I would've implemented it, I would've made the if empty on spawn then add reagents check a global snack thing, instead of only limited to some food items, with an exception for snacks/grown so plants you intend to be empty spawn empty, cause there's a lot of copypaste with this implementation, but I guess it works for the more unique foods that have more than just nutriment.

**Comment by Optimism333** [comment]
> Not how I would've implemented it, I would've made the if empty on spawn then add reagents check a global snack thing, instead of only limited to some food items, with an exception for snacks/grown so plants you intend to be empty spawn empty, cause there's a lot of copypaste with this implementation, but I guess it works for the more unique foods that have more than just nutriment.

Not all food slices are created equal. As you said, some have unique reagents, and many have varying amounts of nutriment in the food object they are sliced from.

If you have a suggestion that reduces copying AND keeps unique reagent values for different slices, please share.

**Comment by pedr69** [comment]
the proper fix in my opinion would be to change how the slicable food is created, probably by passing the nutrient content to the `slice.new()` proc, and if a slice is initialised without the nutrient argument it creates a small amount
https://github.com/vgstation-coders/vgstation13/blob/42d0f9afe96b5d887bc7f1bd83e74bc47b5b569f/code/modules/reagents/reagent_containers/food/snacks.dm#L500

**Comment by SonixApache** [comment]
@Optimism333 spawns like this are bad


@pedr69 
> the proper fix in my opinion would be to change how the slicable food is created, probably by passing the nutrient content to the `slice.new()` proc, and if a slice is initialised without the nutrient argument it creates a small amount(you'd still need to copypaste a ton of code for this, or change all the existing slices to use inheritance)
> 
> https://github.com/vgstation-coders/vgstation13/blob/42d0f9afe96b5d887bc7f1bd83e74bc47b5b569f/code/modules/reagents/reagent_containers/food/snacks.dm#L500

just add a line that say something like ``` if(reagents.total_volume) 
(add_reagent(NUTRIMENT, 3)```
after line 510
https://github.com/vgstation-coders/vgstation13/blob/42d0f9afe96b5d887bc7f1bd83e74bc47b5b569f/code/modules/reagents/reagent_containers/food/snacks.dm#L510
no copypaste, slices fixed

alternatively, to avoid unforeseen consequences add nutriment to anything that doesn't have nutriment in your lunchbox on its post-init.

alternatively alternatively, considering your 'maintain unique reagents' business, spawn a copy of the master item, slice it, move one slice into the lunchbox and delete the rest of the slices - stupid and expensive, but it works

**Comment by Optimism333** [comment]
Gonna switch this back to a draft until I improve the shitcode.

**Comment by Optimism333** [comment]
I removed all the checks from the snacks.dm page, and replaced them with local checks on the lunchbox.dm page per @SonixApache 's post init suggestion.

I added checks to all lunchbox lists that have food slice items as options. When the lunchbox spawns, it will check if its food selection is empty of reagents, and add some nutriment if it is. Mains benefit more from this than snacks/sweets (6 nutriment vs 3 nutriment).

I only added the checks to lunchbox lists that include food slice items, some didn't need any checks: (Discount, Trader).

**Comment by Optimism333** [comment]
Added a requested feature.

Ablative lunchboxes will protect your lunch from lasers! They won't protect you, unfortunately.

https://github.com/vgstation-coders/vgstation13/assets/81007356/dae096e8-6307-46a9-a86a-3682a0475ba4

**Comment by Eneocho** [comment]
> Added a requested feature.
> 
> Ablative lunchboxes will protect your lunch from lasers! They won't protect you, unfortunately.

ablino lunchboxes... can you make it so scientists have a higher chance of rolling the vest lunchbox?

**Comment by Optimism333** [comment]
> > Added a requested feature.
> > Ablative lunchboxes will protect your lunch from lasers! They won't protect you, unfortunately.
> 
> ablino lunchboxes... can you make it so scientists have a higher chance of rolling the vest lunchbox?

Nah, sorry but I don't wanna. I set out to make this pr to add the four missing lunchbox designs, fix the no nutriment problem, and added one little new feature. I'm calling it mission complete on this one.

I'll do additional bugfixes and necessary code improvement, but that's all.

**Comment by Eneocho** [comment]
that's fine, it was just an idea I had, since the ablino scientist is a staple stereotype, them having extra chances for the ablinobox was amusing to me

File: code/game/objects/items/weapons/storage/lunchbox.dm [hunk 2]
```diff
@@ -496,6 +509,39 @@
...
+			desc = "A little plastic lunchbox. This one has a scene decorating the side that depicts four ERT members striking action poses."
+			icon_state = "lunchbox_ert"
+			item_state = "toolbox_lightblue"
+		if(18)
+			name = "Shuttle lunchbox"
+			desc = "A little plastic lunchbox. This one is decorated to look like a little escape shuttle."
+			icon_state = "lunchbox_shuttle"
+			item_state = "toolbox_lightblue2"
+
+/obj/item/weapon/storage/lunchbox/plastic/nt/random/bullet_act(var/obj/item/projectile/P) // Ablative lunchboxes protect lunch from lasers!
```
Comment (Kurfursten) [inline code/game/objects/items/weapons/storage/lunchbox.dm:533]: this is fine for just one gimmick version, but if you ever give alternate features to more lunchboxes you need to make this oop by making them subtypes

**Pull request merged**